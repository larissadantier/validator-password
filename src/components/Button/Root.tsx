import { ComponentProps } from "react";

import styles from './button.module.scss';

function Root({ children, className, ...props }: ComponentProps<'button'>) {
  return (
    <button data-testid="button" className={`${styles['button__container']} ${className}`} {...props}>
      {children}
    </button>
  );
}

export default Root;