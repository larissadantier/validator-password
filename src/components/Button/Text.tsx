import { ComponentProps } from "react";

import styles from './button.module.scss';

function Text({ children, ...props }: ComponentProps<'span'>) {
  return (
    <span className={styles['button__text']} {...props}>{children}</span>
  );
}

export default Text;