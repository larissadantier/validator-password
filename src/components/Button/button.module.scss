.button {
  &__container {
    background-color: var(--white);

    border: 1px solid var(--border-color-button);
    border-radius: 5px;

    padding: 0.5rem 0.75rem;

    box-shadow: var(--box-shadow) 0px 2px 8px 0px;

    &:hover:not(:disabled) {
      background-color: var(--box-shadow-hover);
    }

    &:focus-within,
    &:focus-visible {
      border-color: var(--focus-blue);
      box-shadow: var(--focus-blue-shadow) 0px 2px 8px 0px;
    }
  }

  &__text {
    font-size: 1rem;
    font-weight: 500;
  }
}