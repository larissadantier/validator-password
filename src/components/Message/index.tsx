import { IMessageProps } from './interface';
import { getClassNameVariant } from './utilsStyles';

function Message({
  children,
  variant,
  ...props }: IMessageProps) {
  const classNameVariant = getClassNameVariant(variant)

  return (
    <span
      data-testid="message"
      className={classNameVariant}
      {...props}
    >
      {children}</span>
  );
}

export default Message;