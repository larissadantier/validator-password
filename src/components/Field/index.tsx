import Root from './Root';
import Label from './Label';
import Input from "./Input";
import Warn from './Warn';


export const Field = {
  Root,
  Label,
  Input,
  Warn
}