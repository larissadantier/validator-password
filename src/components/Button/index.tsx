import Root from "./Root";
import Text from './Text';

export const Button = {
  Root,
  Text
}