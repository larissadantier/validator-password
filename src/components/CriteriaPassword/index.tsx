import List from './List';
import Item from './Item';
import Title from './Title';
import Text from './Text';

export const CriteriaPassword = {
  List,
  Item,
  Title,
  Text
}