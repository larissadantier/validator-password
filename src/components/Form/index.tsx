import { useFormValidator } from '../../hooks/useFormValidator';
import { useValidatorPassword } from '../../hooks/useValidatorPassword';

import { Field } from '../../components/Field';
import Message from '../Message';
import { Button } from '../Button';
import { CriteriaPassword } from '../CriteriaPassword';

import { validateEmail } from '../../utils/validator/email';

import { passwordCriteria } from './mock/items';

import styles from './form.module.scss';

function Form() {
  const {
    handleSubmit,
    handleInputChange,
    formData,
    loading,
    responseStatus,
    isFormValid,
  } = useFormValidator();

  const { validatorPasswordState } = useValidatorPassword(formData.password)

  const isValidPassword = Object.values(validatorPasswordState).every(validator => validator === true)

  const messagePassword = isValidPassword ? 'Senha válida' : 'Senha inválida'

  return (
    <form className={styles['form__container']} onSubmit={handleSubmit}>
      <section className={styles['form__content']}>
        <header>
          <h1>Valide a sua senha</h1>
          <p className={styles['form__description']}>de acordo com os critérios de avaliação</p>
        </header>

        <div className={styles['form__fields']}>
          <Field.Root>
            <Field.Label htmlFor='name'>Nome</Field.Label>
            <Field.Input
              name='name'
              id="name"
              required
              value={formData.name}
              placeholder='Ex: Larissa Dantier'
              onChange={handleInputChange}
            />
            {formData.name === '' && <Field.Warn>Preencha o campo de nome</Field.Warn>}
          </Field.Root>

          <Field.Root>
            <Field.Label htmlFor='email'>Email</Field.Label>
            <Field.Input
              name='email'
              id="email"
              type='email'
              required
              value={formData.email}
              placeholder='Ex: lari@gmail.com'
              onChange={handleInputChange}
            />

            {formData.email === '' && <Field.Warn>Preencha o campo de e-mail</Field.Warn>}
            {formData.email && !validateEmail(formData.email) && <Field.Warn>E-mail inválido</Field.Warn>}
          </Field.Root>

          <Field.Root>
            <Field.Label htmlFor='password'>Senha</Field.Label>
            <Field.Input
              name='password'
              id="password"
              required
              value={formData.password}
              placeholder='Ex: 122345'
              type='text'
              inputMode='numeric'
              pattern='[0-9]+'
              maxLength={6}
              onChange={handleInputChange}
            />
            {formData.password === '' && <Field.Warn>Preencha o campo de senha</Field.Warn>}
          </Field.Root>
        </div>

        {formData.password && (
          <CriteriaPassword.List>
            <CriteriaPassword.Title valid={isValidPassword}>{messagePassword}</CriteriaPassword.Title>

            {!isValidPassword && passwordCriteria.map(item => (
              <CriteriaPassword.Item key={item.id}>
                <CriteriaPassword.Text valid={validatorPasswordState[item.type]}>
                  {item.text}
                </CriteriaPassword.Text>
              </CriteriaPassword.Item>
            ))}
          </CriteriaPassword.List>
        )}
      </section>

      <footer className={styles['form__footer']}>
        {responseStatus && (
          <Message variant={responseStatus.variant}>{responseStatus.message}</Message>
        )}

        <Button.Root
          className={styles['form__button']}
          type='submit'
          disabled={loading || !isFormValid || !isValidPassword}
        >
          <Button.Text>Enviar</Button.Text>
        </Button.Root>
      </footer>
    </form>
  );
}

export default Form;