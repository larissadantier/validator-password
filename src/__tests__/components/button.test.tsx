import { expect, describe, it, vi } from 'vitest';
import { fireEvent, render } from '@testing-library/react';

import { Button } from '../../components/Button';

const buttonTestId = 'button';

describe('Component Button', () => {
  it('should be able render the button', () => {
    const { getByTestId } = render(
      <Button.Root>
        <Button.Text>Send</Button.Text>
      </Button.Root>
    )

    expect(getByTestId(buttonTestId)).toBeInTheDocument();
  })

  it('should be able user to click in button', () => {
    const handleClick = vi.fn();

    const { getByTestId } = render(
      <Button.Root onClick={handleClick}>
        <Button.Text>Send</Button.Text>
      </Button.Root>
    );

    fireEvent.click(getByTestId(buttonTestId));

    expect(handleClick).toHaveBeenCalled();
  })

  it('should be able user to focus', () => {
    const { getByTestId } = render(
      <Button.Root>
        <Button.Text>Send</Button.Text>
      </Button.Root>
    );

    const button = getByTestId(buttonTestId);

    button.focus();

    expect(button).toHaveFocus();
  })

  it('should be able not receive click when disabled', () => {
    const handleClick = vi.fn();

    const { getByTestId } = render(
      <Button.Root onClick={handleClick} disabled>
        <Button.Text>Send</Button.Text>
      </Button.Root>
    );
    const button = getByTestId('button');

    fireEvent.click(button);

    expect(handleClick).not.toHaveBeenCalled();
  });

  it('should match snapshot', () => {
    const { asFragment } = render(
      <Button.Root>
        <Button.Text>Send</Button.Text>
      </Button.Root>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should match snapshot of disabled', () => {
    const { asFragment } = render(
      <Button.Root disabled>
        <Button.Text>Send</Button.Text>
      </Button.Root>
    );

    expect(asFragment()).toMatchSnapshot();
  });
})