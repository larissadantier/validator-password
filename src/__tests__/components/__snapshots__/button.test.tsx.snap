// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Component Button > should match snapshot 1`] = `
<DocumentFragment>
  <button
    class="_button__container_f6b8e0 undefined"
    data-testid="button"
  >
    <span
      class="_button__text_f6b8e0"
    >
      Send
    </span>
  </button>
</DocumentFragment>
`;

exports[`Component Button > should match snapshot of disabled 1`] = `
<DocumentFragment>
  <button
    class="_button__container_f6b8e0 undefined"
    data-testid="button"
    disabled=""
  >
    <span
      class="_button__text_f6b8e0"
    >
      Send
    </span>
  </button>
</DocumentFragment>
`;
